x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/NegMapper.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.planner;
+
+import org.elasticsearch.compute.operator.EvalOperator;
+import org.elasticsearch.xpack.ql.expression.predicate.operator.arithmetic.Neg;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+abstract class NegMapper extends EvalMapper.ExpressionMapper<Neg> {
+    static final EvalMapper.ExpressionMapper<?> NEG_MAPPER = new NegMapper(
+        org.elasticsearch.xpack.esql.expression.predicate.operator.arithmetic.NegIntsEvaluator::new,
+        org.elasticsearch.xpack.esql.expression.predicate.operator.arithmetic.NegLongsEvaluator::new,
+        org.elasticsearch.xpack.esql.expression.predicate.operator.arithmetic.NegDoublesEvaluator::new
+    ) {
+    };
+
+    private final BiFunction<Source, EvalOperator.ExpressionEvaluator, EvalOperator.ExpressionEvaluator> ints;
+
+    private final BiFunction<Source, EvalOperator.ExpressionEvaluator, EvalOperator.ExpressionEvaluator> longs;
+    private final Function<EvalOperator.ExpressionEvaluator, EvalOperator.ExpressionEvaluator> doubles;
+
+    private NegMapper(
+        BiFunction<Source, EvalOperator.ExpressionEvaluator, EvalOperator.ExpressionEvaluator> ints,
+        BiFunction<Source, EvalOperator.ExpressionEvaluator, EvalOperator.ExpressionEvaluator> longs,
+        Function<EvalOperator.ExpressionEvaluator, EvalOperator.ExpressionEvaluator> doubles
+    ) {
+        this.ints = ints;
+        this.longs = longs;
+        this.doubles = doubles;
+    }
+
+    @Override
+    protected final Supplier<EvalOperator.ExpressionEvaluator> map(Neg neg, Layout layout) {
+        DataType type = neg.dataType();
+        if (type.isNumeric()) {
+            var childEvaluator = EvalMapper.toEvaluator(neg.field(), layout).get();
+
+            // Unsigned longs are unsupported by choice; negating them does not make much sense without converting to long;
+            // that would require implicit conversions to long also for expressions like `x-y` for consistency.
+            if (type == DataTypes.INTEGER) {
+                return () -> ints.apply(neg.source(), childEvaluator);
+            }
+            if (type == DataTypes.LONG) {
+                return () -> longs.apply(neg.source(), childEvaluator);
+            }
+            if (type == DataTypes.DOUBLE) {
+                return () -> doubles.apply(childEvaluator);
+            }
+        }
+        throw new AssertionError("arithmetic negation operator with unsupported data type");
```

Comment:
Nit: might be worth logging the fraudulent type.

Reply:
To clarify: since the `AssertionError` seems to kill the server with the given error message, I'll add the data type to the error message - is that what you had in mind? (I didn't see usage of an actual logger in any of the other mapper classes.)

Reply:
Further yet - never throw an error; an exception is enough.
AssertionError are used by unit tests.

Reply:
Agree!

Actually, `AssertionError` is used in multiple mappers (and I only used it to be consistent with the behavior of e.g. `ArithmeticMapper`.

To keep the PR from becoming noisy, how about I do a follow-up PR that kicks out all `AssertionError`s from the mappers and replaces them with an appropriate exception, @costin ?

Reply:
Went with @luigidellaquila 's suggestion of using `UnsupportedOperationException` for now, will remove `AssertionError` from other mappers in a follow-up PR.